AbstractOptions isset issue, #7286

rms230 opened the pull request:
When calling isset() on an object variable that is a class extending AbstractOptions an exception is thrown.

An example useage of this is found in the EventManagerAwareTrait where the trait checks a class variable isset, if the class that is using the EventManagerAwareTrait extends AbstractOptions an exception is thrown.

This pull request fixes AbstractOptions by checking that the method exists and that the value is not equal to null, thus stopping the exception being thrown.

This pull request also fixes issue #7286

rms230 committed:
When checking a variable is set on a class that extends AbstractOptions that doesn't have the variable, an exception is thrown.

rms230 committed:
Fixes for issue #7286

Martin-P commented:
Perhaps you can split up and create a separate PR for each issue? I see a fix for `Zend\StdLib\AbstractOptions` and also for a completely unrelated issue about `Zend\Form\View\Helper\FormCheckbox`.

malukenho commented on library/Zend/Form/View/Helper/FormCheckbox.php:
```diff
@@ -60,6 +60,7 @@ public function render(ElementInterface $element)
             $hiddenAttributes = array(
                 'name'  => $attributes['name'],
                 'value' => $element->getUncheckedValue(),
+				'disabled' => isset($attributes['disabled']) ? $attributes['disabled'] : false,
```
Use `spaces` instead of `tabs`

malukenho commented on tests/ZendTest/Form/View/Helper/FormCheckboxTest.php:
```diff
@@ -83,6 +83,20 @@ public function testDoesNotThrowExceptionIfNameIsZero()
+
+
+		$markup = $this->helper->__invoke($element);
+		$this->assertRegexp('#type="checkbox".*?(disabled="disabled")#', $markup);
```
Remove unnecessary spaces and use spaces too.

malukenho commented on tests/ZendTest/Stdlib/OptionsTest.php:
```diff
@@ -112,4 +113,36 @@ public function testExceptionMessageContainsActualUsedSetter()
+			isset($options->foo);
+		} catch (\BadMethodCallException $ex) {}
+
+		$this->assertNull($ex, 'Unexpected BadMethodCallException');
```
Use **phpunit** assert for exception `$this->setExpectedException('BadMethodCallException')`

weierophinney commented on library/Zend/Stdlib/AbstractOptions.php:
```diff
@@ -147,7 +147,8 @@ public function __get($key)
     public function __isset($key)
     {
-        return null !== $this->__get($key);
+		$getter = 'get' . str_replace('_', '', $key);
```
Spaces instead of tabs, please.

weierophinney commented:
@rms230 : Please read our [contributing guide](https://github.com/zendframework/zf2/blob/master/README-GIT.md#working-on-zend-framework), particularly the section linked on branching for patches. The pull request here combines two separate features, which should be evaluated separately; this is due to doing both on your master branch. Also make sure you run php-cs-fixer (which is installed by default when you run `composer install`) to correct any CS issues you have; this will make it simpler for us to merge.

You can create a new branch from your master branch, revert the unwanted commits (the stuff around the form components), and create a new pull request from there. The changes *do* look good; I just want to make sure they come in cleanly.

Thanks!

Commit:
Update FormCheckbox.php

Commit:
Update FormCheckbox.php

Commit:
Update FormCheckboxTest.php

Commit:
Update AbstractOptions.php

Commit:
Update FormCheckbox.php

Commit:
Update OptionsTest.php

The test checks that the exception was not thrown, this checks that future implementations do not revert changes were made.

Checking for an exception thrown would cause the tests to fail. How would I go about doing this with the standards?

weierophinney commented:
@rms230 If you want this included in 2.4.0, you have until tomorrow, 19 March 2015, to resubmit as a discrete patch against develop as outlined in my previous comment.

Martin-P commented on tests/ZendTest/Form/View/Helper/FormCheckboxTest.php:
```diff
@@ -83,6 +83,16 @@ public function testDoesNotThrowExceptionIfNameIsZero()
+        $element->setUseHiddenElement(true);
+        $element->setAttribute('disabled', true);
+        $markup = $this->helper->__invoke($element);
+        $this->assertRegexp('#type="checkbox".*?(disabled="disabled")#', $markup);
```
Just tested this and this test already passes against current master.

Martin-P commented:
@weierophinney I created a separate PR for the checkbox issue and also updated the unittest so it fails against current master and passes with the provided fix from @rms230 See #7341.

weierophinney commented:
Closed in favor of #7341

Martin-P commented:
@weierophinney Please see PR #7342 for the second part of this PR about the `AbstractOptions` issue.
